packages/library-sync/src/commonMain/kotlin/io/realm/kotlin/mongodb/Functions.kt

```diff
@@ -0,0 +1,103 @@
+/*
+ * Copyright 2022 Realm Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.realm.kotlin.mongodb
+
+import io.realm.kotlin.mongodb.exceptions.FunctionExecutionException
+import io.realm.kotlin.mongodb.internal.BsonEncoder
+import kotlinx.serialization.DeserializationStrategy
+import kotlinx.serialization.KSerializer
+import kotlinx.serialization.serializer
+import kotlinx.serialization.serializerOrNull
+import org.mongodb.kbson.BsonArray
+import org.mongodb.kbson.BsonDocument
+import org.mongodb.kbson.BsonValue
+import kotlin.reflect.typeOf
+
+/**
+ * A Functions manager to call remote Atlas Functions for the associated Atlas App services Application.
+ *
+ * Due to the serialization engine [does not support third-party libraries yet](https://github.com/Kotlin/kotlinx.serialization/blob/master/docs/formats.md), there are some
+ * limitations in what types can be used as arguments and return types:
+ *
+ * - Primitives, Bson, lists, and maps are valid argument types.
+ * - Results can only be deserialized to primitives or Bson types.
+ *
+ * @see [User.functions]
+ */
+public interface Functions {
+    /**
+     * App where the Atlas Functions would be executed on.
+     */
+    public val app: App
+
+    /**
+     * User that to authenticate the Atlas Function calls.
+     */
+    public val user: User
+
+    /**
+     * Invokes an Atlas function.
+     *
+     * Due to the serialization engine [does not support third-party libraries yet](https://github.com/Kotlin/kotlinx.serialization/blob/master/docs/formats.md), there are some
+     * limitations in what types can be used as arguments and return types:
+     *
+     * - Primitives, Bson, lists, and maps are valid argument types.
+     * - Results can only be deserialized to primitives or Bson types.
+     *
+     * The Bson implementations for arrays or maps are [BsonArray] and [BsonDocument], and they can be
+     * used as valid return types.
+     *
+     * @param name Name of the function to call.
+     * @param args Arguments to the function.
+     * @param deserializationStrategy Deserialization strategy for decoding the results.
+     * @param T The type for the functions response.
+     * @return Result of the function.
+     *
+     * @throws FunctionExecutionException if the request failed in some way.
+     */
+    public suspend fun <T : Any?> call(
+        name: String,
+        deserializationStrategy: DeserializationStrategy<T>,
```

Comment:
The serializers are picked automatically from the decoder being used, by resolving it with the type of the argument. 

To be able to deserialize the return value, we need to resolve the serializer from the type, thus the reified parameter. We could pass the return value class and then later resolve it, but it is easier to do in the inline function.

The Json decoder API works like that, you can decode using an [inline-reified](https://github.com/Kotlin/kotlinx.serialization/blob/16a85df254f4f1e317554eb61ee1fbe914800aa4/formats/json/commonMain/src/kotlinx/serialization/json/Json.kt#L209) function or[ manually pass the serializer.](https://github.com/Kotlin/kotlinx.serialization/blob/16a85df254f4f1e317554eb61ee1fbe914800aa4/formats/json/commonMain/src/kotlinx/serialization/json/Json.kt#L114)

Reply:
I had a more detailed look at this and I don't feel this is the way to go. We are currently setting up the `DeserializationStrategy`-based API just to be able to identify the type but are not using the actual deserializers at all. I also don't think that the users should be bothered with doing conversion to EJSON but only be bothered with converting objects to BSON-values. This functionality of converting objects should probably go into the BSON library just as the codec support of the Java BSON library. 

Given how easy it is to do a `@RequiresOptIn`-annotation (see https://github.com/realm/realm-kotlin/commit/dc83ae7c1e689cfdd0ce9dfc2630473637814d07) I think we should just keep the most basic functions API with something like
```
public suspend fun call(name: String, arg: BsonValue): BsonValue
```
Then we can do a very basic temporary `BsonConverter` with the default primitive type conversions until we have a more solid idea of the overall API should look like. This would also be symmetric around yielding Basically what I outlined in https://github.com/realm/realm-kotlin/commit/41a9744538b642c6b871b9167b51494ff5536cbf ... Didn't rip out all the `DeserializationStrategy` so also didn't rip out the `BsonEncoder` just yet. Anyway ... up for discussing the details.

Reply:
The kserializer API for writing third-party encoders is still experimental. 

We designed the API to support in the short future a non-experimental serializer with reduced type support (current implementation) and a full blown experimental one.

To do so I took it from the starting point of what the final API should look like, I took the Json encoder as a reference, as the Bson encoder is going to have the same api except for the method names.

I see the point of your concern on the API wise. I think that we might strip the BsonEncoder thing from this PR, and leave it in the bare minimum, and address it in a future PR.